src/charm.py

```diff
@@ -53,8 +54,19 @@ def _on_install(self, event) -> None:
         # Prevent the default cluster creation.
         self._cluster.inhibit_default_cluster_creation()
 
-        # Install the PostgreSQL packages.
-        self._install_apt_packages(event, ["postgresql"])
+        # Install the PostgreSQL and Patroni requirements packages.
+        self._install_apt_packages(event, ["postgresql", "python3-pip", "python3-psycopg2"])
+
+        try:
+            resource_path = self.model.resources.fetch("patroni")
+        except ModelError as e:
+            logger.error(f"missing patroni resource {str(e)}")
+            self.unit.status = BlockedStatus("Missing 'patroni' resource")
+            return
+
+        # Build Patroni package path with raft dependency and install it.
+        patroni_package_path = f"{str(resource_path)}[raft]"
+        self._install_pip_packages([patroni_package_path])
```

Comment:
Same here,  on failure to install `pip` packages the code goes into `BlockedStatus` and then executes then and then goes into `WaitingStatus`

To prevent it from going into waiting status, maybe something like this (or similar) would do
```
# need to make _install_pip_packages return a boolean
install_successful = self._install_pip_packages([patroni_package_path]) 
if not install_successful:
      return
```

Reply:
Thanks for catching it, Mia!